walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,236 @@
+   * cannot be part of the potential meeting time.
+   */
+  private List<TimeRange> getAllBusyTimeRanges(
+      Collection<Event> events, Collection<String> mandatoryAttendees) {
+    List<TimeRange> busyTimes = new ArrayList<TimeRange>();
+
+    for (Event event : events) {
+      Collection<String> eventAttendees = event.getAttendees();
+
+      for (String attendee : eventAttendees) {
+        // If the current event's attendee list contains a mandatory attendee,
+        // adds this event's time range to the set of all busy times,
+        // after checking for overlapping time ranges.
+        if (mandatoryAttendees.contains(attendee)) {
+          TimeRange currEventTimeRange = event.getWhen();
+
+          busyTimes = getUpdatedBusyTimeRanges(currEventTimeRange, busyTimes);
+
+          // If any one of the mandatory attendees is in this event's guest list,
+          // this event's time range needs to be considered busy.
+          // Breaks out of the for-each loop because there is no need to look
+          // at the rest of the event's guest list.
+          break;
+        }
+      }
+    }
+    return busyTimes;
+  }
+
+  /**
+   * Compares a given event's time slot with all the existing busy time slots. Merges the current
+   * event time slot with a busy time slot if they overlap. Adds the current event's time slot to
+   * the list of busy slots in the end.
+   *
+   * @param current event's time range; a collection of busy time slots
+   * @return an updated list of busy times after overlaps are handled
+   */
+  private List<TimeRange> getUpdatedBusyTimeRanges(
+      TimeRange currEventTimeRange, Collection<TimeRange> originalBusyTimes) {
+    boolean isOverlapped = false;
+
+    List<TimeRange> updatedBusyTimes = new ArrayList<TimeRange>();
+
+    for (TimeRange busyTimeRange : originalBusyTimes) {
+      if (currEventTimeRange.overlaps(busyTimeRange)) {
+        TimeRange combinedTimeRange = getCombinedTimeRange(currEventTimeRange, busyTimeRange);
+        updatedBusyTimes.add(combinedTimeRange);
+        isOverlapped = true;
+        currEventTimeRange = combinedTimeRange;
+      } else {
+        updatedBusyTimes.add(busyTimeRange);
+      }
+    }
+
+    // If the current event doesn't overlap with any existing busy time range,
+    // add the current event's time range to the set of all busy time slots.
+    if (!isOverlapped) {
+      updatedBusyTimes.add(currEventTimeRange);
+    }
+
+    return updatedBusyTimes;
+  }
+
+  /** Creates a new overall combined time range for two time ranges that overlap. */
+  private static TimeRange getCombinedTimeRange(TimeRange a, TimeRange b) {
+    int aStart = a.start();
+    int aEnd = a.end();
+    int bStart = b.start();
+    int bEnd = b.end();
+
+    TimeRange combinedTimeRange =
+        TimeRange.fromStartEnd(
+            /* start= */ Math.min(aStart, bStart),
+            /* end= */ Math.max(aEnd, bEnd),
+            /* inclusive= */ false);
```

Comment:
Doesn't this `inclusive` value have to depend on `a` and `b`? Like if say `bEnd > aEnd` and `b.inclusive() == true`, then you should pass `true` here I think?

Reply:
I think, based on the TimeRange class that is given to us, there isn't really a way to find the "inclusive" attribute of a TimeRange. From reading the class and the provided test cases, it seems that the default is exclusive for end time. The exception is if the end time is END_OF_DAY, then this time range should be inclusive of the end. 
In my method, I changed it so that, if the later end time between a and b is the END_OF_DAY, then I set the inclusive flag to true, otherwise the newly created time range is always exclusive of the end time.

Reply:
I see ya you're right. Looking at the code in `TimeRange` they just subtract `1` from `end` if it's exclusive.